### PR TITLE
sql: add support for map constructors

### PR DESCRIPTION
**Pull request by benesch**
Add support for long desired map constructors, akin to list and array constructors. Maps can now be constructed with the `MAP` literal:

    MAP['a' => 1, 'b' => 2]

Or from a subquery:

    MAP(VALUES ('a', 1), ('b', 2))

See the documentation within for details.

### Motivation

  * This PR adds a feature that has not yet been specified.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Add a `MAP` expression that can construct a map from a list of key–value pairs or a subquery.

### CODE REVIEW

**Review by ggevay — COMMENTED** [review]
Looks good from a Compute perspective. I wrote some ideas for additional tests.

**Review by maddyblue — APPROVED** [review]
Code LGTM. SQL syntax LGTM.

File: doc/user/content/releases/v0.100.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+---
+title: "Materialize v0.100"
+date: 2024-05-22
+released: false
+---
+
+## v0.100
```
Comment (morsapaes) [inline doc/user/content/releases/v0.100.md:7]: :rip:

**Review by morsapaes — APPROVED** [review]

**Review by sploiselle — APPROVED** [review]
I added a few cases that demonstrate that inner expressions of keys and values work. https://github.com/MaterializeInc/materialize/pull/27006/commits/3f06efdbbd0db4acc298673598e3ec0e97ed92fb

**Review by benesch — COMMENTED** [review]
Thanks! Good call to test that.

**Comment by nrainer-materialize** [comment]
https://github.com/MaterializeInc/materialize/pull/27062 will add further tests.

**Pull request merged**